daml-lf/kv-support/src/main/scala/com/daml/lf/kv/transactions/TransactionConversions.scala

```diff
@@ -92,9 +95,157 @@ object TransactionConversions {
+      rawTransaction: RawTransaction
+  ): Either[ConversionError, Set[ContractIdOrKey]] =
+    Try(TransactionOuterClass.Transaction.parseFrom(rawTransaction.byteString)) match {
+      case Failure(throwable) =>
+        Left(ConversionError.ParseError(throwable.getMessage))
+      case Success(transaction) =>
+        TransactionCoder
+          .decodeVersion(transaction.getVersion)
+          .flatMap { txVersion =>
+            transaction.getNodesList.asScala
+              .foldLeft[Either[ValueCoder.DecodeError, Set[ContractIdOrKey]]](Right(Set.empty)) {
+                case (Right(contractIdsOrKeys), node) =>
+                  TransactionCoder.decodeNodeVersion(txVersion, node).flatMap { nodeVersion =>
+                    node.getNodeTypeCase match {
+                      case NodeTypeCase.ROLLBACK =>
+                        // Nodes under rollback will potentially produce outputs such as divulgence.
+                        // Actual outputs must be a subset of, or the same as, computed outputs and
+                        // we currently relax this check by widening the latter set, treating a node the
+                        // same regardless of whether it was under a rollback node or not.
+                        // Computed outputs that are not actual outputs can be safely trimmed: examples
+                        // are transient contracts and, now, potentially also outputs from nodes under
+                        // rollback.
+                        Right(contractIdsOrKeys)
+
+                      case NodeTypeCase.CREATE =>
+                        val protoCreate = node.getCreate
+                        for {
+                          newContractIdsOrKeys <- TransactionCoder
+                            .nodeKey(nodeVersion, protoCreate)
+                            .map {
+                              case Some(key) => contractIdsOrKeys + ContractIdOrKey.Key(key)
+                              case None => contractIdsOrKeys
+                            }
+                          contractId <- ValueCoder.CidDecoder
+                            .decode(protoCreate.getContractIdStruct)
+                        } yield newContractIdsOrKeys + ContractIdOrKey.Id(contractId)
+
+                      case NodeTypeCase.EXERCISE =>
+                        val protoExercise = node.getExercise
+                        for {
+                          newContractIdsOrKeys <- TransactionCoder
+                            .nodeKey(nodeVersion, protoExercise)
+                            .map {
+                              case Some(key) => contractIdsOrKeys + ContractIdOrKey.Key(key)
+                              case None => contractIdsOrKeys
+                            }
+                          contractId <- ValueCoder.CidDecoder
+                            .decode(protoExercise.getContractIdStruct)
+                        } yield newContractIdsOrKeys + ContractIdOrKey.Id(contractId)
```

Comment:
I don’t understand the intended semantics here. Looking at the usesite I thought this was about outputs. But here you are including contract ids of non-consuming exercises which doesn’t seem like something that should be an output.

Reply:
Oh looking below I guess it’s because it can cause divulgence like the fetch.

Reply:
Looks like it's the case. But TBH, I don't know, and it would be great to get more explanation on that. Maybe @miklos-da knows the details?

Reply:
@hubert-da I'd suggest figuring out who made the last changes to this section and asking that person — unfortunately, I don't have context either.